pkg/controller/nodeipam/ipam/multicidrset/multi_cidr_set.go

```diff
@@ -0,0 +1,365 @@
+	// Stores a mapping of node name to association status.
+	AssociatedNodes map[string]bool
+	// Terminating is used to identify whether ClusterCIDR has been marked for termination.
+	Terminating bool
+}
+
+const (
+	// The subnet mask size cannot be greater than 16 more than the cluster mask size
+	// TODO: https://github.com/kubernetes/kubernetes/issues/44918
+	// clusterSubnetMaxDiff limited to 16 due to the uncompressed bitmap.
+	// Due to this limitation the subnet mask for IPv6 cluster cidr needs to be >= 48
+	// as default mask size for IPv6 is 64.
+	clusterSubnetMaxDiff = 16
+	// halfIPv6Len is the half of the IPv6 length.
+	halfIPv6Len = net.IPv6len / 2
+)
+
+// CIDRRangeNoCIDRsRemainingErr is an error type used to denote there is no more
+// space to allocate CIDR ranges from the given CIDR.
+type CIDRRangeNoCIDRsRemainingErr struct {
+	// CIDR represents the CIDR which is exhausted.
+	CIDR string
+}
+
+func (err *CIDRRangeNoCIDRsRemainingErr) Error() string {
+	return fmt.Sprintf("CIDR allocation failed; there are no remaining CIDRs left to allocate in the range %s", err.CIDR)
+}
+
+// CIDRSetSubNetTooBigErr is an error type to denote that subnet mask size is too
+// big compared to the CIDR mask size.
+type CIDRSetSubNetTooBigErr struct {
+	cidr            string
+	subnetMaskSize  int
+	clusterMaskSize int
+}
+
+func (err *CIDRSetSubNetTooBigErr) Error() string {
+	return fmt.Sprintf("Creation of New CIDR Set failed for %s. "+
+		"PerNodeMaskSize %d is too big for CIDR Mask %d, Maximum difference allowed "+
+		"is %d", err.cidr, err.subnetMaskSize, err.clusterMaskSize, clusterSubnetMaxDiff)
+}
+
+// NewMultiCIDRSet creates a new MultiCIDRSet.
+func NewMultiCIDRSet(cidrConfig *net.IPNet, perNodeHostBits int) (*MultiCIDRSet, error) {
+	clusterMask := cidrConfig.Mask
+	clusterMaskSize, bits := clusterMask.Size()
+
+	var subNetMaskSize int
+	switch /*v4 or v6*/ {
+	case cidrConfig.IP.To4() != nil:
+		{
+			subNetMaskSize = 32 - perNodeHostBits
+		}
+	case cidrConfig.IP.To16() != nil:
+		{
+			subNetMaskSize = 128 - perNodeHostBits
+		}
+	}
+
+	if (cidrConfig.IP.To4() == nil) && (subNetMaskSize-clusterMaskSize > clusterSubnetMaxDiff) {
+		return nil, &CIDRSetSubNetTooBigErr{
+			cidr:            cidrConfig.String(),
+			subnetMaskSize:  subNetMaskSize,
+			clusterMaskSize: clusterMaskSize,
+		}
+	}
+
+	// Register MultiCIDRSet metrics.
+	registerCidrsetMetrics()
+
+	return &MultiCIDRSet{
+		ClusterCIDR:      cidrConfig,
+		nodeMask:         net.CIDRMask(subNetMaskSize, bits),
+		clusterMaskSize:  clusterMaskSize,
+		MaxCIDRs:         1 << uint32(subNetMaskSize-clusterMaskSize),
+		NodeMaskSize:     subNetMaskSize,
+		Label:            cidrConfig.String(),
+		AllocatedCIDRMap: make(map[string]bool, 0),
+	}, nil
+}
+
+func (s *MultiCIDRSet) indexToCIDRBlock(index int) (*net.IPNet, error) {
+	var ip []byte
+	switch /*v4 or v6*/ {
+	case s.ClusterCIDR.IP.To4() != nil:
+		{
```

Comment:
you don't need to add braces around a switch case

Reply:
probably worth pull these conversion routines out into a func  itself and making sure it is unit tested